# Assignment...

#### 1. Write a Python script to display the various Date Time formats -
a) Current date and time
b) Current year
c) Month of year
d) Week number of the year
e) Weekday of the week
f) Day of year
g) Day of the month
h) Day of week

In [33]:
# a) Current date and time
from datetime import datetime
today=datetime.now()
print("Today's date: ",today)

formatted_datetime = today.strftime("%Y-%m-%d %H:%M:%S")
print("Formatted Date and Time:", formatted_datetime)

# b) Current year
print("\nCurrent year:", today.year)

# c) Month of year
print("\nCurrent month:", today.strftime("%B"))

# d) Week number of the year
print("\nCurrent week of the year:", today.strftime("%U"))

# e) Weekday of the week
print("\nWeekday of the week:", today.strftime("%A"))

# f) Day of year
print("\nDay of year:", today.strftime("%j"))

# g) Day of the month
print("\nDay of the month:", today.strftime("%d"))

# h) Day of week
print("\nDay of week:", today.strftime("%w"))

Today's date:  2024-02-05 11:35:08.265530
Formatted Date and Time: 2024-02-05 11:35:08

Current year: 2024

Current month: February

Current week of the year: 05

Weekday of the week: Monday

Day of year: 036

Day of the month: 05

Day of week: 1


#### 2.Write a Python program to subtract five days from the current date.
Sample Date :
Current Date : 2015-06-22

In [44]:
from datetime import date, timedelta
today_date=date.today()
print(today_date-timedelta(days=5))
print(f"\n{today_date+timedelta(days=5)}")

2024-01-31

2024-02-10


#### 3.Write a Python program to calculate the number of days between two date times.

In [53]:
start_date_str = "2024-01-01 05:00:00"
end_date_str = "2024-02-01 12:30:00"

x=datetime.strptime(start_date_str, "%Y-%m-%d %H:%M:%S")
y=datetime.strptime(end_date_str, "%Y-%m-%d %H:%M:%S")
                    
difference = y - x

print(difference)

31 days, 7:30:00


#### 4.Write a Python program to read an entire text file.

In [69]:
file_path="example.txt"
content="""Cricket is a popular sport played and watched by millions of fans around the world.
It is a bat-and-ball game played between two teams, each consisting of eleven players.
The game is played on a circular field with a rectangular 22-yard-long pitch in the center.
The primary objective is for the batting team to score runs by hitting the ball and running between two sets of wickets.
The fielding team aims to dismiss the batsmen or restrict the runs scored.
Cricket has different formats, including Test matches, One Day Internationals (ODIs), and Twenty20 (T20) matches.
The sport has a rich history, with international competitions like the ICC Cricket World Cup and the Ashes series being highly prestigious.
Legendary players like Sachin Tendulkar, Sir Donald Bradman, and Brian Lara have left a significant impact on the game.
Cricket has a diverse fan base, and major cricketing nations include India, Australia, England, Pakistan, and South Africa.
The game's unique terminology, such as "bowled," "lbw" (leg before wicket), and "century," adds to its charm and complexity."""
with open (file_path,'w') as file:
    file.write(content)
    
print(file)

with open (file_path,'r') as file:
    contents=file.read()
    
print(f"\n{contents}")

<_io.TextIOWrapper name='example.txt' mode='w' encoding='cp1252'>

Cricket is a popular sport played and watched by millions of fans around the world.
It is a bat-and-ball game played between two teams, each consisting of eleven players.
The game is played on a circular field with a rectangular 22-yard-long pitch in the center.
The primary objective is for the batting team to score runs by hitting the ball and running between two sets of wickets.
The fielding team aims to dismiss the batsmen or restrict the runs scored.
Cricket has different formats, including Test matches, One Day Internationals (ODIs), and Twenty20 (T20) matches.
The sport has a rich history, with international competitions like the ICC Cricket World Cup and the Ashes series being highly prestigious.
Legendary players like Sachin Tendulkar, Sir Donald Bradman, and Brian Lara have left a significant impact on the game.
Cricket has a diverse fan base, and major cricketing nations include India, Australia, England, Paki

#### 5.Write a Python program to read first n lines of a file.

In [88]:
n=int(input())
with open("example.txt", "r") as file:
    for line in range(n):
        lines=file.readline()
        if not lines:
            break
        print(lines)

3
Cricket is a popular sport played and watched by millions of fans around the world.

It is a bat-and-ball game played between two teams, each consisting of eleven players.

The game is played on a circular field with a rectangular 22-yard-long pitch in the center.



#### 6.Write a Python program to read last n lines of a file.

In [101]:
# n=int(input())
# with open("example.txt", "r") as file:
#     lines=file.readlines()
#     startindex=len(lines)-n
#     for line in lines[startindex:]:
#         print(line)

n=int(input())
with open("example.txt", "r") as file:
    for line in file.readlines()[-n:]:
            print(line)

3
Legendary players like Sachin Tendulkar, Sir Donald Bradman, and Brian Lara have left a significant impact on the game.

Cricket has a diverse fan base, and major cricketing nations include India, Australia, England, Pakistan, and South Africa.

The game's unique terminology, such as "bowled," "lbw" (leg before wicket), and "century," adds to its charm and complexity.


#### 7.Write a Python program to get the file size of a plain file.

In [104]:
import os
size_of_file = os.path.getsize(file_path)
print(size_of_file)

1090


#### 8.Write a Python program that takes a text file as input and returns the number of words of a given text file.
Note: Some words can be separated by a comma with no space.

In [107]:
file_path="example1.txt"
content="Hi I am Kushagra"
with open (file_path,'w') as file:
    file.write(content)

with open("example1.txt", "r") as file:    
    text = file.read()
    words = text.replace(",", " ").split()
    num_words = len(words)
    print(num_words)

4


#### 9.Write a Python program to generate 26 text files named A.txt, B.txt, and so on up to Z.txt.

In [109]:
import string
for i in string.ascii_uppercase:
    file_path=f"{i}.txt"
    with open (file_path,'w') as file:
        file.write(f"The name of file is {i}")

#### 10.Write a Python program to read each row from a given csv file and print a list of strings.

In [151]:
import csv
file_path="idcard.csv"
csv_data = [["Name", "Age", "City"],
    ["John", 25, "New York"],
    ["Alice", 30, "London"],
    ["Bob", 22, "Paris"],
]

with open(file_path, "w", newline='') as csv_file:
    writer = csv.writer(csv_file)
    for row in csv_data:
        writer.writerow(row)

with open(file_path, "r") as csv_file:
    reader = csv.reader(csv_file)
    for row in reader:
        print(row)

['Name', 'Age', 'City']
['John', '25', 'New York']
['Alice', '30', 'London']
['Bob', '22', 'Paris']


#### 11.Write a Python program to read a given CSV file having tab delimiter.

In [152]:
import csv
file_path="idcard.csv"

with open(file_path, "r") as csv_file:
    reader = csv.reader(csv_file,delimiter='\t')
    for row in reader:
        print(row)

['Name,Age,City']
['John,25,New York']
['Alice,30,London']
['Bob,22,Paris']


#### 12.Write a Python program that reads a CSV file and remove initial spaces, quotes around each entry and the delimiter.

In [153]:
import csv
file_path="idcard.csv"
with open(file_path, "r") as csv_file:
    reader = csv.reader(csv_file)
    cleaned_row = [[entry.strip().strip('"') for entry in row]for row in reader]
    for row in cleaned_row:
        print(row)

['Name', 'Age', 'City']
['John', '25', 'New York']
['Alice', '30', 'London']
['Bob', '22', 'Paris']


#### 13.Write a Python program to read specific columns of a given CSV file and print the content of the columns.

In [164]:
import csv
file_path="idcard.csv"
column=[0,1]
with open(file_path, "r") as csv_file:
    reader = csv.reader(csv_file)        

    for row in reader:
        selected_data = [row[index] for index in column]
        print(selected_data)

['Name', 'Age']
['John', '25']
['Alice', '30']
['Bob', '22']


# Assignment...

#### 1.Reading from CSV file

In [166]:
file_path="stocks.csv"
with open(file_path, 'r') as csv_file:
    reader=csv.reader(csv_file)
    for row in reader:
        print(row)

['Company Name', 'Price', 'Earnings Per Share', ' Book Value']
['Reliance', '1467', '66', '653']
['Tata Steel', '391', '89', '572']
['Infosys', '650', '35', '147']
['Axis Bank', '739', '19', '263']
['Bajaj Finance', '4044', '69', '341']


#### 2.Writing to CSV file
writerow()
writerows()

In [179]:
file_path="example3.csv"
with open(file_path, "w", newline='') as csv_file:
    writer = csv.writer(csv_file)
    single_row = ["Adarsh", 19, "Jhansla"]
    writer.writerow(single_row)

    multiple_rows = [
        ["Harsh", 21, "Chandigarh"],
        ["Kus", 22, "Mohali"],
        ["Hims", 20, "Rajpura"]
    ]
    writer.writerows(multiple_rows)

#### 3.print the min, max and avg score of given score.csv file

In [197]:
file_path="scores.csv"
with open(file_path, "r") as csv_file:
    reader=csv.reader(csv_file)
    min_value=float('inf')
    max_value=float('-inf')
    total_value=0
    counter=0
    for row in reader:
        score=float(row[1])
        min_value=min(min_value,score)
        max_value=max(max_value,score)
        total_value+=score
        counter+=1
        
    if counter == 0:
        print("No scores found in the CSV file.")
    else:
        print(f"min_value is: {min_value}")
        print(f"max_value is: {max_value}")
        print(f"avg_value is: {total_value/counter}")

min_value is: 5.0
max_value is: 140.0
avg_value is: 78.92857142857143


#### 4.Appending To a File in Python
i- Append-adds at last
ii-Write-Overwrites

In [222]:
file_path="example4.csv"
with open (file_path,'a') as csv_file:
    csv_file.write("This is appended content.\n")
    
with open (file_path,'w') as csv_file:
    csv_file.write("This is being overwritten.")

#### 5.Reading Data From an XML File
i-Finding Tags 
ii-Extracting from tags

In [289]:
from bs4 import BeautifulSoup

with open("content.xml", "r") as file:
    xml_data = file.read()

soup = BeautifulSoup(xml_data, "xml")
print(soup)

row_tags = soup.find_all("shape")
print()
print(row_tags)

<?xml version="1.0" encoding="utf-8"?>
<data xmlns="http://example.com">
<row>
<shape>square</shape>
<degrees>360</degrees>
<sides>4.0</sides>
<firstname>Krish</firstname>
</row>
<row>
<shape>circle</shape>
<degrees>360</degrees>
<sides/>
<firstname/>
</row>
<row>
<shape>triangle</shape>
<degrees>180</degrees>
<sides>3.0</sides>
<firstname/>
</row>
</data>

[<shape>square</shape>, <shape>circle</shape>, <shape>triangle</shape>]


#### 6.Reading the data inside the xml file to a variable under the name data

In [260]:
from bs4 import BeautifulSoup

with open("content.xml", "r") as file:
    xml_data = file.read()
    
soup = BeautifulSoup(xml_data, 'xml')

data = soup.find_all('row')
# print(data)
# print()
for row in data:
    print(row)

<row>
<shape>square</shape>
<degrees>360</degrees>
<sides>4.0</sides>
<firstname>Krish</firstname>
</row>
<row>
<shape>circle</shape>
<degrees>360</degrees>
<sides/>
<firstname/>
</row>
<row>
<shape>triangle</shape>
<degrees>180</degrees>
<sides>3.0</sides>
<firstname/>
</row>


#### 7.Writing XML Files

In [307]:
from bs4 import BeautifulSoup

with open("work.xml",'r') as f:
    data=f.read()
    
soup=BeautifulSoup(data,"xml")

for tag in soup.find_all('shape'):
    tag.string = 'circle'
    
print(soup)

with open("modified_work.xml", 'w') as f:
    f.write(soup.prettify())

<?xml version="1.0" encoding="utf-8"?>
<data xmlns="http://example.com">
<row>
<shape>circle</shape>
<degrees>360</degrees>
<sides>4.0</sides>
<firstname>Krish</firstname>
</row>
<row>
<shape>circle</shape>
<degrees>360</degrees>
<sides/>
<firstname/>
</row>
<row>
<shape>circle</shape>
<degrees>180</degrees>
<sides>3.0</sides>
<firstname/>
</row>
</data>


#### 8.Writing List to Files in Python
Using write()
Using writelines()
Using String Join Along with “with open” syntax

In [294]:
my_list = ['Sachin Tendulkar', 'Virat Kohli', 'Ricky Ponting', 'Brian Lara', 'Wasim Akram']

file_path = 'example5.txt'
with open(file_path, 'w') as file:
    for item in my_list:
        file.write(f"{item}\n")
print(f"List written to {file_path} using write() successfully.")

my_list = [['Sachin Tendulkar', 'Virat Kohli', 'Ricky Ponting', 'Brian Lara', 'Wasim Akram'],['Kapil Dev', 'Anil Kumble', 'Javagal Srinath', 'Harbhajan Singh', 'Ravichandran Ashwin']]
file_path = 'example6.txt'
with open(file_path, 'w') as file:
    for item in my_list:
        file.writelines(f"{item}\n")
print(f"List written to {file_path} using write() successfully.")

my_list = [['Sachin Tendulkar', 'Virat Kohli', 'Ricky Ponting', 'Brian Lara', 'Wasim Akram'],['Kapil Dev', 'Anil Kumble', 'Javagal Srinath', 'Harbhajan Singh', 'Ravichandran Ashwin']]
file_path = 'example7.txt'
with open(file_path, 'w') as file:
    for item in my_list:
        file.write('\n'.join(item))
print(f"List written to {file_path} using string join successfully.")

List written to example5.txt using write() successfully.
List written to example6.txt using write() successfully.
List written to example7.txt using string join successfully.


#### 9.[poem.txt]contains famous poem "Road not taken" by poet Robert Frost. You have to read this file in your python program and find out words with maximum occurance.

In [243]:
with open("poem.txt",'r') as f:
    data=f.read()
    cleaned_string=data.replace(",", "").replace(".", "").replace(":", "").replace(";", "").replace("-", "")
    
    word_count={}
    words=cleaned_string.split()
    for word in words:
        word_count[word] = word_count.get(word, 0) + 1

    max_count=0
    print("Word Count Dictionary:")
    for word, count in word_count.items():
        max_count=max(max_count,count)
        
    for word, count in word_count.items():
        if(max_count==count):
            print(word)


Word Count Dictionary:
I
the


#### 10.[stocks.csv]contains stock price, earnings per share and book value. You are writing a stock market application that will process this file and create a new file
with financial metrics such as pe ratio and price to book ratio. These are calculated as,
```
pe ratio = price / earnings per share
price to book ratio = price / book value
```


Your input format (stocks.csv) is,

|Company Name|Price|Earnings Per Share|Book Value|
|-------|----------|-------|----------|
|Reliance|1467|66|653|
|Tata Steel|391|89|572|

Output.csv should look like this,

|Company Name|PE Ratio|PB Ratio|
|-------|----------|-------|
|Reliance|22.23|2.25|
|Tata Steel|4.39|0.68|

In [288]:
import csv
with open("stocks.csv","r") as f:
    reader=csv.DictReader(f)
    
    output=[]
    output_headers=['Company Name','PE Ratio','PB Ratio']
    
    for row in reader:
        company_name=row['Company Name']
        price=float(row['Price'])
        earning_per_share=float(row['Earnings Per Share'])
        book_value=float(row['Book Value'])
        
        pb_ratio=round(price/earning_per_share,2)
        pe_ratio=round(price/book_value,2)
        
        output.append({'Company Name':company_name, 'PE Ratio':pe_ratio, 'PB Ratio':pb_ratio})
        
with open("Output.csv","w",newline="") as f:
    writer=csv.DictWriter(f, fieldnames=output_headers)
    writer.writeheader()
    writer.writerows(output)